legion/sdk/legion/sdk/clients/edi.py

```diff
@@ -331,3 +321,210 @@ def info(self):
+
+
+class AsyncRemoteEdiClient(RemoteEdiClient):
+
+    async def query(self, url_template: str, payload: Mapping[Any, Any] = None, action: str = 'GET'):
+        """
+        Perform query to EDI server
+
+        :param url_template: url template from legion.const.api
+        :param payload: payload (will be converted to JSON) or None
+        :param action: HTTP method (GET, POST, PUT, DELETE)
+        :return: dict[str, any] -- response content
+        """
+        response = await self._request(url_template, payload, action)
+        return self._handle_query_response(response, payload)
+
+    async def _request(self, url_template: str,
+                       payload: Mapping[Any, Any] = None,
+                       action: str = 'GET',
+                       stream: bool = False,
+                       timeout: Optional[int] = None,
+                       limit_stack: bool = False,
+                       client: Optional[httpx.AsyncClient] = None):
+        """
+        Make HTTP request
+
+        :param url_template: target URL
+        :param payload: data to be placed in body of request
+        :param action: request action, e.g. get / post / delete
+        :param stream: use stream mode or not
+        :param timeout: custom timeout in seconds (overrides default). 0 for disabling
+        :param limit_stack: do not start refreshing token if it is possible
+        :return: :py:class:`requests.Response` -- response
+        """
+
+        assert not stream or stream and client, "You must pass AsyncClient for streaming response"
+
+        request = self._build_request(httpx.AsyncRequest, url_template, payload, action, stream)
+
+        connection_timeout = self._get_conn_timeout(timeout)
+        left_retries = self._get_left_retries()
+
+        raised_exception = None
+        while left_retries > 0:
+            try:
+                LOGGER.debug('Requesting {}'.format(request.url))
+
+                if client:
+                    response = await client.send(request, timeout=connection_timeout, stream=stream)
+                else:
+                    async with httpx.AsyncClient() as client:
+                        response = await client.send(request, timeout=connection_timeout, stream=stream)
+
+                resp_text = response.text if not stream else '<stream>'
+                LOGGER.debug('Status code: "{}", Response: "{}"'.format(response.status_code, resp_text))
+
+            except (httpx.exceptions.ConnectTimeout, socket.gaierror) as exception:
+                LOGGER.error('Failed to connect to {}: {}. Retrying'.format(self._base_url, exception))
+                raised_exception = exception
+            else:
+                LOGGER.debug('Got response. Breaking')
+                break
+            left_retries -= 1
+        else:
+            raise EDIConnectionException('Can not reach {}'.format(self._base_url)) from raised_exception
+
+        if self._login_required(response):
```

Comment:
All our security functions, which interacts with a Keycloak, are synchronous. Is it expected?
`legion/jyputerlab/handlers/oauth2.py` is too.

Reply:
Yes, I supposed to add async features iteratively. It allows not to rewrite all codebase at once. Therefore I optimized only SDK methods that affect jupyter performance